stacker/config/__init__.py

```diff
@@ -37,3 +37,164 @@ def parse_config(raw_config, environment=None):
     buff.seek(0)
     config = yaml.load(buff)
     return config
+
+
+class Dumper(yaml.Dumper):
+    pass
+
+
+class Loader(yaml.Loader):
+    pass
+
+
+def parse(raw, validate=True):
+    """Parse and validate a stacker config.
+
+    Args:
+        raw (str): the raw stacker configuration string in yaml format.
+        validate (bool): whether or not to validate the parsed config.
+
+    Returns:
+        :class:`Config`: the parse (and validated) stacker config.
+
+    """
+
+    config = yaml.load(raw, Loader=Loader) or Config()
+
+    if validate:
+        config.validate()
+
+    return config
+
+
+def dump(config):
+    """Dumps a stacker Config object as yaml.
+
+    Args:
+        config (:class:`Config`): the stacker Config object.
+
+    Returns:
+        str: the yaml formatted stacker Config.
+
+    """
+    return yaml.dump(config, Dumper=Dumper)
+
+
+class Object(yaml.YAMLObject):
+    yaml_loader = Loader
+    yaml_dumper = Dumper
+
+    def __init__(self, **kwargs):
+        self.__setstate__(kwargs)
+
+    def needs(self, key):
+        if not self.__dict__.get(key):
+            raise exceptions.MissingConfig(key)
+
+
+class Config(Object):
+    """This is the Python representation of a stacker config file.
+    """
+
+    yaml_tag = "!Config"
+
+    def __setstate__(self, state):
+        self.namespace = state.get("namespace")
+        self.namespace_delimiter = state.get("namespace_delimiter")
+        self.stacker_bucket = state.get("stacker_bucket")
+        self.sys_path = state.get("sys_path")
+        self.package_sources = state.get("package_sources")
+        self.pre_build = state.get("pre_build")
+        self.post_build = state.get("post_build")
+        self.pre_destroy = state.get("pre_destroy")
+        self.post_destroy = state.get("post_destroy")
+        self.tags = state.get("tags")
+        self.mappings = state.get("mappings")
+        self.lookups = state.get("lookups")
+        self.stacks = state.get("stacks")
+
+    def validate(self):
+        self.needs("namespace")
+        self.needs("stacks")
+
+        if self.stacks:
+            for stack in self.stacks:
+                stack.validate()
+
+        if self.package_sources:
+            self.package_sources.validate()
+
+        phases = [
+                self.pre_build, self.post_build,
+                self.pre_destroy, self.post_destroy]
+        for hooks in phases:
+            if hooks:
+                for hook in hooks:
+                    hook.validate()
+
+
+class Stack(Object):
+    yaml_tag = "!Stack"
+
+    def __setstate__(self, state):
+        self.name = state.get("name")
+        self.class_path = state.get("class_path")
+        self.requires = state.get("requires")
+        self.locked = state.get("locked", False)
+        self.enabled = state.get("enabled", True)
+        self.variables = state.get("variables")
```

Comment:
This could probably use more structure, so that lookups are parsed out at this point in time. That'd help a lot with cleaning up the crazy regex's we currently have for doing this, making a lot of the internals for resolving lookups simpler.